Add support for readonly members in structs (#673)
dotnet/csharpstandard: standard/structs.md

- Instance field declarations for a struct are not permitted to include variable initializers ([§16.4.8](structs.md#1648-field-initializers)).
- A struct is not permitted to declare a parameterless instance constructor ([§16.4.9](structs.md#1649-constructors)).
- A struct is not permitted to declare a finalizer.

> - Some kinds of members in structs are permitted to have the modifier `readonly` while that is not generally permitted for those same member kinds in classes.

Reviewer: While I can appreciate the desire not to repeat things, I suspect it's worth being more specific about which members can have `readonly` in structs. The "some" here feels really woolly.

Reviewer: This would be addressed by my suggestion below, assuming I haven't missed anything.

> - Some kinds of members in structs are permitted to have the modifier `readonly` while that is not generally permitted for those same member kinds in classes.

Reviewer: ```suggestion
- Event declarations, property declarations, property accessors, indexer declarations, and method declarations are permitted to have the modifier `readonly` while that is not generally permitted for those same member kinds in classes.
```

> *Note*: This access restriction means that constructors in structs containing automatically implemented properties often need an explicit constructor initializer where they would not otherwise need one, to satisfy the requirement of all fields being definitely assigned before any function member is invoked or the constructor returns. *end note*

A *property_declaration* ([§15.7.1](classes.md#1571-general)) for an instance property in a *struct_declaration* may contain the *property_modifier* `readonly`. However, a static property shall not contain that modifier.

> It is a compile-time error to attempt to modify the state of an instance struct variable via a readonly property declared in that struct.

Reviewer: "attempt" feels more human than spec-like to me. What *rule* are we stating here, and can we do so more precisely?

Reviewer 2: Agree with @jskeet and additionally we have some repetition here which I think should be factored out as it is for classes: under classes modifiers are generally described in subclauses under *§15.3 Class members*.

That same pattern can be repeated here with the current content of *§16.3 Struct members* becoming *§16.3.1 General* and adding *§16.3.2 The readonly modifier*.

In §16.3.2 the rules for readonly members can be given (not proposed text):

- They cannot mutate the state of the struct instance
- They may mutate the static state of the struct
- And the (subjectively) ugly one, a readonly member may call a mutating member (of the same instance only?) but any mutations made must not be observable after the call. This could in terms of the compiler strategy of “defensive copies”, see #1053, or not.

This would provide a single place for defining what a readonly member. The additional per-member differences could follow in §16.3 subclauses or remain as they are in §16.4 subclauses.

Reviewer: My review 2025-06-02: I think this is still work to do, although I'd be fine with creating an issue for it and merging this.

Reviewer 3: Logged as #1339

> It is a compile-time error to attempt to modify the state of an instance struct variable via a readonly property declared in that struct.

Reviewer 2: From [MS Learn](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-8.0/readonly-instance-members):

> Readonly can be applied to property accessors to indicate that this will not be mutated in the accessor. The following examples have readonly setters because those accessors modify the state of member field, but do not modify the value of that member field.
> ```
> public readonly int Prop1
> {
>     get
>     {
>         return this._store["Prop1"];
>     }
>     set
>     {
>         this._store["Prop1"] = value;
>     }
> }
> ```

This example demonstrates a subtly that might easily be missed from the statement on line 589:

> It is a compile-time error to attempt to modify the state of an instance struct variable via a readonly property declared in that struct.

An example like (maybe exactly) this one should be included in the Standard with explanatory text to cover:

- The “indirect” state of a readonly only struct can be modified; and
- That the somewhat oxymoronic `read-only set` makes sense in C#

Unless there are such explanations already here somewhere…

Reviewer 3: I like the idea of an example.  The "read only set that changes the state of a member" is a clear extension of what C# already permits:

```csharp
public struct S
{
    private List<int> sequence;

    public IList<int> Sequence
    {
        get
        {
            sequence ??= new List<int>();
            return sequence;
        }
    }
}
```

Callers can modify the returned sequence. The property above can't be marked `readonly` until a future version of C# allows field initializers in struct (10, if I recall correctly).

Reviewer 3: Comment added on #1339

It is a compile-time error for an automatically implemented property having a `readonly` modifier, to also have a `set` accessor.

It is a compile-time error for an automatically implemented property in a `readonly` struct to have a `set` accessor.

> An automatically implemented property declared inside a `readonly` struct need not have a `readonly` modifier, as its `get` accessor is implicitly assumed to be readonly.

Reviewer 4: Roslyn also treats auto property getter of non-readonly structs as implicitly readonly. Should this be intended in spec?

Reviewer 5: This behaviour of Roslyn avoids copying the non-`readonly` `struct` value when reading an automatically implemented property of an `in` parameter:

```csharp
struct S {
    int I { get; set; }
    int J => 42;

    static int M(in S s) {
        int i = s.I; // does not copy s
        int j = s.J; // defensive copy, as in ((S)s).J
        return i + j;
    }
}
```

However, I don't think a conforming C# program can detect whether this copying happens, because the `get` accessor of `S.I` does not contain any user code that could mutate the original instance via an alias and check whether the mutation also affected `this`.  (For C# implementations that target the CLR, it can be still detected by disassembling the IL code.)  Therefore, I think the C# standard does not have to specify it.

Reviewer 3: I think the standard can remain quiet on this behavior. A conforming compiler shouldn't be *required* to add the readonly modifier in this instance.

Reviewer: My review 2025-06-02: I'm happy to merge without a change here.

It is a compile-time error for a property to have a `readonly` modifier on both the `get` and the `set` accessors.

It is a compile-time error to have a `readonly` modifier on a property itself as well as on either of its `get` and `set` accessors.

If the `get` accessor has a `readonly` modifier, the `set` shall exist and shall not have that modifier.

### §cands-diffs-methods Methods

A *method_declaration* ([§15.6.1](classes.md#1561-general)) for an instance method in a *struct_declaration* may contain the *method_modifier* `readonly`. However, a static method shall not contain that modifier.

> It is a compile-time error to attempt to modify the state of an instance struct variable via a readonly method declared in that struct.

Reviewer: Another case of "attempt".

Reviewer: My review 2025-06-02: Add an issue rather than trying to fix this before merging? (Maybe a single issue for line 509 and this.)

Reviewer 3: See #1339

> Although a readonly method may call a sibling, non-readonly method, or property or indexer get accessor, doing so results in the creation of an implicit copy of `this` as a defensive measure.

Reviewer 3: We haven't defined "sibling" anywhere in the spec. I suggest replacing it with "may call a ... **declared in the same `struct` declaration**" (emphasis mine for suggested text.

Reviewer 5: The wording "declared in the same `struct` declaration" would incorrectly cause a copy here:

```csharp
struct S {
    void M(in S s) {}
    readonly void N(S s) {
        s.M(this); // must not make an implicit copy of `this`
    }
}
```

Reviewer 2: While I wrote a short while ago in #1053 that I wasn’t intending to get into the defensive copy issue at this time re-reading comments like the above makes me think we have little choice.

Last year I suggested the [common aspects of readonly need to be defined in single place](https://github.com/dotnet/csharpstandard/pull/673#discussion_r1510548259) and that including covering that readonly members may call mutating members but any mutations must not be visible afterwards – i.e. specify *what* not *how*. Do this and we should find wording things like the above easier.

And I think we'll need to define *sibling*, at least for this topic if not for the wider Standard.

Reviewer: My review 2025-06-02: I suggest we file an issue for sibling, but not block this PR on it.

Reviewer 3: See #1339

A readonly method may call a sibling property or indexer set accessor that is readonly. If a sibling member’s accessor is not explicitly or implicitly readonly, a compile-error occurs.

All *method_declaration*s of a partial method shall have a `readonly` modifier, or none of them shall have it.

> ### §cands-diffs-indexers Indexers

Reviewer: So just to check, that means an indexer *can* have a `readonly set` (so long as it doesn't have a `readonly get` as well)?

I guess if it's modifying the state of something referenced via the struct, that's reasonable. Odd though.

Reviewer 3: Yes. That's exactly why. @jnm2 has a proposal to allow this in more places: https://github.com/dotnet/csharplang/blob/main/proposals/readonly-setter-calls-on-non-variables.md  (Note: not currently in the working set)

An *indexer_declaration* ([§15.9](classes.md#159-indexers)) for an instance indexer in a *struct_declaration* may contain the *indexer_modifier* `readonly`.

> It is a compile-time error to attempt to modify the state of an instance struct variable via a readonly indexer declared in that struct.

Reviewer: Another case of "attempt"

Reviewer: (Probably to be included in the same issue as for line 509 etc.)

Reviewer 3: See #1339

It is a compile-time error to have a `readonly` modifier on both the `get` and the `set` accessors.

It is a compile-time error to have a `readonly` modifier on an indexer itself as well as on either of its `get` or `set` accessors.

> If the `get` accessor has a `readonly` modifier, the `set` shall exist and shall not have that modifier.

Reviewer: ```suggestion
If the `get` accessor has a `readonly` modifier, the `set` accessor shall exist and shall not have that modifier.
```

Reviewer 3: The resolution here should follow the same pattern chosen for the similar restriction on line 515 regarding properties.

Reviewer 3: The same fix was applied here and for properties.

> *Note*: This access restriction means that constructors in structs containing automatically implemented properties often need an explicit constructor initializer where they would not otherwise need one, to satisfy the requirement of all fields being definitely assigned before any function member is invoked or the constructor returns. *end note*

> A *property_declaration* ([§14.7.1](classes.md#1471-general)) for an instance property in a *struct_declaration* may contain the *property_modifier* `readonly`. However, a static property shall not contain that modifier.

Reviewer 5: Does _"However, a static property shall not contain that modifier"_ also cover the case where the `readonly` modifier is on an accessor of the static property?

```
struct S {
    static int P {
        readonly get => 0;
        set {}
    }
}
```

It is a compile-time error to attempt to modify the state of an instance struct variable via a readonly property declared in that struct.

It is a compile-time error for an automatically implemented property having a `readonly` modifier, to also have a `set` accessor.

It is a compile-time error for an automatically implemented property in a `readonly` struct to have a `set` accessor.

An automatically implemented property declared inside a `readonly` struct need not have a `readonly` modifier, as its `get` accessor is implicitly assumed to be readonly.

It is a compile-time error for a property to have a `readonly` modifier on both the `get` and the `set` accessors.

It is a compile-time error to have a `readonly` modifier on a property itself as well as on either of its `get` and `set` accessors.

> If the `get` accessor has a `readonly` modifier, the `set` shall exist and shall not have that modifier.

Reviewer 5: Should disallow this as well:

```csharp
struct S {
    int P { // error CS8664: 'S.P': 'readonly' can only be used on accessors if the property or indexer has both a get and a set accessor
        readonly set { }
    }
}
```

Perhaps:

> If an accessor has a `readonly` modifier, the property shall also have an accessor that does not have that modifier.

Reviewer 5: Or merge with the earlier rule

> It is a compile-time error for a property to have a `readonly` modifier on both the `get` and the `set` accessors.

resulting in:

> It is a compile-time error for a property to have a `readonly` modifier on all of its accessors.  *Note:* To correct the error, move the modifier from the accessors to the property itself.  *end note*

Reviewer 2: Agree, assuming CS8664 is the intended semantics and not a Roslyn quirk.

IIRC conformance allows *additional* members other than `get` & `set` so “all of its accessors” would be invalid in such a case. I'll offer the following:

```suggestion
If the `get` or `set` accessor has a `readonly` modifier then a sibling `set` or `get` accessor shall exist and shall not have that modifier.
```

Could use “corresponding” rather than “sibling”

Reviewer 5: Assuming the implementation allows a property to have a nonstandard `increment` accessor in addition to the standard `get` and `set`:

```csharp
struct S {
    private int i;
    public int I {
        readonly get => i;
        increment => ++i;
    }
}
```

I think "all of its accessors" would be a good rule here.  If the developer wants all accessors to be readonly, then they should make the whole property readonly, regardless of whether the accessors are standard or nonstandard.  And if the developer wants the `get` accessor to be readonly but wants some nonstandard accessor not to be readonly, then the standard should not require them to make the whole property readonly, as that change would also affect the nonstandard accessor (or require contortions in the specification of the extension).

Reviewer 3: Well, the grammar currently only allows `get` and `set` accessors. (See 15.7.3). Given that, I'd prefer the simpler "all accessors", but I'm not too concerned one way or the other

Reviewer: My review 2025-06-02: We should do something here, and I don't think we need to worry about non-standard accessors. I'd be fine with the "It is a compile-time error for a property to have a readonly modifier on all of its accessors. Note: To correct the error, move the modifier from the accessors to the property itself. end note" suggestion.

Reviewer 3: I made an edit that follows Jon's suggestion.

### §cands-diffs-methods Methods

A *method_declaration* ([§14.6.1](classes.md#1461-general)) for an instance method in a *struct_declaration* may contain the *method_modifier* `readonly`. However, a static method shall not contain that modifier.

It is a compile-time error to attempt to modify the state of an instance struct variable via a readonly method declared in that struct.

> Although a readonly method may call a sibling, non-readonly method, or property or indexer get accessor, doing so results in the creation of an implicit copy of `this` as a defensive measure.

Reviewer 5: OK, demo:

```csharp
struct S {
    int P => 0; // not readonly
    
    int this[string s] => 0; // not readonly

    readonly void M() {
        _ = P; // warning CS8656: Call to non-readonly member 'S.P.get' from a 'readonly' member results in an implicit copy of 'this'.
        _ = this[""]; // warning CS8656: Call to non-readonly member 'S.this[string].get' from a 'readonly' member results in an implicit copy of 'this'.
        N(); // warning CS8656: Call to non-readonly member 'S.N()' from a 'readonly' member results in an implicit copy of 'this'.
    }

    void N() {}
}
```

The standard does not require these warnings and I think that's OK.

Reviewer 5: Another demo, also OK:

```csharp
struct S {
    private static int r;
    
    public ref int P {
        get => ref r; // not readonly
    }
    
    readonly void M() {
        this.P = 0; // warning CS8656: Call to non-readonly member 'S.P.get' from a 'readonly' member results in an implicit copy of 'this'.
    }
}
```

`this.P` is at the left side of the assignment, but this calls the `get` accessor and so the wording is OK.

Reviewer 2: Agree in both cases, these warnings are up to a compiler. The first example in particular does however raise another potential issue – see #1053.

Reviewer 3: I made a note in #1053: I think we should defer this until we spec out `ref readonly` parameter in C# 12.

Reviewer: My review 2025-06-02: Happy to defer this.

> A readonly method may call a sibling property or indexer set accessor that is readonly. If a sibling member’s accessor is not explicitly or implicitly readonly, a compile-error occurs.

Reviewer 5: Does "sibling" specifically mean an instance member of `this`?  The term is not used with that meaning anywhere else in the specification.  It is not clear that it excludes static members, or instance members of the same type but used on a different instance.

Reviewer 2: Another thing that could be defined in §16.3.x as per suggestion on line 509

Reviewer 3: See #1339

> A readonly method may call a sibling property or indexer set accessor that is readonly. If a sibling member’s accessor is not explicitly or implicitly readonly, a compile-error occurs.

Reviewer 5: Is there already a rule against `C.N` here:

```csharp
struct S {
    public int P {
        set {} // not readonly
    }
    
    public int this[string s] {
        set {} // not readonly
    }

    readonly void M() {
        this.P = 0; // error CS1604: Cannot assign to 'this.P' because it is read-only
        this[""] = 0; // error CS1604: Cannot assign to 'this[""]' because it is read-only
    }
}

class C {
    static void N(in S that) {
        that.P = 0; // error CS8332: Cannot assign to a member of variable 'in S' or use it as the right hand side of a ref assignment because it is a readonly variable
        that[""] = 0; // error CS8332: Cannot assign to a member of variable 'in S' or use it as the right hand side of a ref assignment because it is a readonly variable
    }
}
```

Not clear to me why these are separate diagnostic codes.  Can the C# standard cover both `S.M` and `C.N` with the same rule?

Reviewer 2: The separate diagnostic codes (and text) are not mandated, that is a compiler choice – these messages are just (in the view of the compiler writer at least, YMMV) better than “error: LHS is read-only”

All *method_declaration*s of a partial method shall have a `readonly` modifier, or none of them shall have it.

### §cands-diffs-indexers Indexers

An *indexer_declaration* ([§14.9](classes.md#149-indexers)) for an instance indexer in a *struct_declaration* may contain the *indexer_modifier* `readonly`.

It is a compile-time error to attempt to modify the state of an instance struct variable via a readonly indexer declared in that struct.

It is a compile-time error to have a `readonly` modifier on both the `get` and the `set` accessors.

It is a compile-time error to have a `readonly` modifier on an indexer itself as well as on either of its `get` or `set` accessors.

If the `get` accessor has a `readonly` modifier, the `set` shall exist and shall not have that modifier.

### §cands-diffs-events Events

> An *event_declaration* ([§14.8.1](classes.md#1481-general)) for an instance, non-field-like event in a *struct_declaration* may contain the *event_modifier* `readonly`. However, a static event shall not contain that modifier.

Reviewer 5: OK; accessors of an event cannot have the `readonly` modifier, but the grammar does not allow any other modifiers on them either, so it does not need to be restated here.
